test/flush_worker_test.cc

```diff
@@ -0,0 +1,149 @@
+
+#include "external/envoy/source/common/common/random_generator.h"
+#include "external/envoy/source/common/event/dispatcher_impl.h"
+#include "external/envoy/source/common/runtime/runtime_impl.h"
+#include "external/envoy/source/common/stats/isolated_store_impl.h"
+#include "external/envoy/test/mocks/api/mocks.h"
+#include "external/envoy/test/mocks/local_info/mocks.h"
+#include "external/envoy/test/mocks/protobuf/mocks.h"
+#include "external/envoy/test/mocks/stats/mocks.h"
+#include "external/envoy/test/mocks/thread_local/mocks.h"
+#include "external/envoy/test/test_common/simulated_time_system.h"
+
+#include "client/flush_worker_impl.h"
+
+#include "gtest/gtest.h"
+
+namespace Nighthawk {
+namespace Client {
+
+using namespace testing;
+
+class FlushWorkerTest : public Test {
+public:
+  FlushWorkerTest()
+      : thread_factory_(Envoy::Thread::threadFactoryForTest()),
+        dispatcher_(new NiceMock<Envoy::Event::MockDispatcher>()) {
+    loader_ = std::make_unique<Envoy::Runtime::ScopedLoaderSingleton>(
+        Envoy::Runtime::LoaderPtr{new Envoy::Runtime::LoaderImpl(
+            *dispatcher_, tls_, {}, local_info_, store_, rand_, validation_visitor_, mock_api_)});
+    sink_ = new StrictMock<Envoy::Stats::MockSink>();
+    stats_sinks_.emplace_back(sink_);
+
+    EXPECT_CALL(mock_api_, threadFactory()).WillRepeatedly(ReturnRef(thread_factory_));
+    // mock_api_ takes the ownership of dispatcher_.
+    EXPECT_CALL(mock_api_, allocateDispatcher_(_, _)).WillOnce(Return(dispatcher_));
+  }
+
+  // We set up emulating timer firing and moving simulated time forward in
+  // simulateTimerloop() below.
+  void setupDispatcherTimerEmulation() {
+    timer_ = new NiceMock<Envoy::Event::MockTimer>();
+    EXPECT_CALL(*dispatcher_, createTimer_(_)).WillOnce(Invoke([&](Envoy::Event::TimerCb cb) {
+      timer_cb_ = std::move(cb);
+      return timer_;
+    }));
+    EXPECT_CALL(*timer_, enableTimer(_, _))
+        .WillRepeatedly(Invoke([&](const std::chrono::microseconds,
+                                   const Envoy::ScopeTrackedObject*) { timer_set_ = true; }));
+    EXPECT_CALL(*timer_, disableTimer()).WillOnce(Invoke([&]() { timer_set_ = false; }));
+    EXPECT_CALL(*dispatcher_, exit()).WillOnce(Invoke([&]() {
+      auto guard = std::make_unique<Envoy::Thread::LockGuard>(lock_);
+      stopped_ = true;
+    }));
+  }
+
+  void expectDispatcherRun() {
+    EXPECT_CALL(*dispatcher_, run(_))
+        .WillOnce(Invoke([&](Envoy::Event::DispatcherImpl::RunType type) {
+          // The first dispatcher run is in WorkerImpl::start().
+          ASSERT_EQ(Envoy::Event::DispatcherImpl::RunType::NonBlock, type);
+        }))
+        .WillOnce(Invoke([&](Envoy::Event::DispatcherImpl::RunType type) {
+          // The second dispatcher run is in FlushWorkerImpl::work().
+          ASSERT_EQ(Envoy::Event::DispatcherImpl::RunType::RunUntilExit, type);
+          simulateTimerLoop();
+        }));
+  }
+
+  void simulateTimerLoop() {
```

Comment:
If we already have the existence of MockTimer, why do we need to simulate a timer?

Reply:
It is because the MockTimer does not act like a real timer, it can not periodically trigger a timer callback, so we need to have simulateTimerLoop(). This is following the example in
https://github.com/envoyproxy/nighthawk/blob/dba15a8a604fda23a78f32862ce349f892c4ac0a/test/sequencer_test.cc#L127